dsl/src/main/scala/asobu/dsl/Extractor.scala

```diff
@@ -1,51 +1,105 @@
 package asobu.dsl
 
-import cats.data.{Xor, XorT}
+import cats.data.{Kleisli, Xor, XorT}
 import play.api.mvc.Results._
 import play.api.mvc.{AnyContent, Request, Result}
-import shapeless.ops.hlist.Prepend
-import shapeless.{HList, HNil}
+import shapeless.ops.hlist._
+import shapeless._
 
 import scala.concurrent.Future
 import scala.util.{Failure, Success, Try}
-import cats.data.Xor._
+import ExtractResult._
+import cats.syntax.all._
+import CatsInstances._
+import concurrent.ExecutionContext.Implicits.global
+import cats.sequence._
 
-object Extractor {
+object Extractor extends ExtractorFunctions
 
-  def fromTry[Repr <: HList](f: Request[AnyContent] ⇒ Try[Repr]): Extractor[Repr] = { req ⇒
-    Future.successful {
-      f(req) match {
-        case Success(repr) ⇒ Right(repr)
-        case Failure(ex)   ⇒ Left(BadRequest(ex.getMessage))
-      }
-    }
+trait ExtractorOps {
+  import Extractor._
+
+  implicit class extractorOps[TFrom, Repr <: HList](self: Extractor[TFrom, Repr]) {
+    def and[ThatR <: HList, ResultR <: HList](that: Extractor[TFrom, ThatR])(
+      implicit
+      prepend: Prepend.Aux[Repr, ThatR, ResultR]
+    ): Extractor[TFrom, ResultR] = combine(self, that)
   }
+}
 
-  val empty: Extractor[HNil] = apply(_ ⇒ HNil)
+trait ExtractorFunctions extends ExtractorOps {
+  implicit def fromFunction[TFrom, T](f: TFrom ⇒ ExtractResult[T]): Extractor[TFrom, T] = Kleisli(f)
 
-  def apply[Repr <: HList](f: Request[AnyContent] ⇒ Repr): Extractor[Repr] = fromTry(f andThen (Try[Repr](_)))
+  implicit def fromFunctionXorT[TFrom, T](f: TFrom ⇒ XorTF[T]): Extractor[TFrom, T] = f.andThen(ExtractResult(_))
 
-  def fromEither[Repr <: HList](f: Request[AnyContent] ⇒ Future[Either[Result, Repr]]): Extractor[Repr] = {
-    import scala.concurrent.ExecutionContext.Implicits.global
-    f.andThen(_.map(Xor.fromEither))
-  }
-}
+  def empty[TFrom]: Extractor[TFrom, HNil] = apply(_ ⇒ HNil)
 
-trait ExtractorOps {
-  import CatsInstances._
+  def apply[TFrom, T](f: TFrom ⇒ T): Extractor[TFrom, T] = f map pure
 
-  implicit class extractorOps[Repr <: HList](self: Extractor[Repr]) {
-    def and[ThatR <: HList, ResultR <: HList](that: Extractor[ThatR])(
-      implicit
-      prepend: Prepend.Aux[Repr, ThatR, ResultR]
-    ): Extractor[ResultR] = { req ⇒
+  def decorate[A, B, C, D](before: A ⇒ B)(extractor: Extractor[B, C])(after: C ⇒ D): Extractor[A, D] =
+    extractor.contramap(before).map(after)
 
-      (for {
-        eitherRepr ← XorT(self(req))
-        eitherThatR ← XorT(that(req))
-      } yield eitherRepr ++ eitherThatR).value
+  /**
+   * extractor from a list of functions
+   * e.g. from(a = (_:Request[AnyContent]).headers("aKay"))
+   */
+  object composeF extends shapeless.RecordArgs {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Object name does not match the regular expression '^[A-Z][A-Za-z0-9]*$'](https://www.codacy.com/app/iHeartRadio/asobu/file/2296169975/issues/source?bid=3037259&fileBranchId=3178877#l46)